File: app/src/main/java/ch/unstable/ost/StationListAdapter.java
package ch.unstable.ost;

import android.content.Context;
import android.os.Handler;
import android.support.annotation.DrawableRes;
import android.support.annotation.MainThread;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import ch.unstable.ost.api.model.Location;
import ch.unstable.ost.theme.ThemeHelper;

import static android.support.v7.widget.RecyclerView.NO_POSITION;

class StationListAdapter extends RecyclerView.Adapter<StationListAdapter.ViewHolder> {

    private final Handler mHandler;
    private final
    @DrawableRes
    int trainIcon;
    private final
    @DrawableRes
    int busIcon;
    private Location[] mLocations = new Location[0];
    @Nullable
    private OnStationClickListener mOnStationClickListener;
    private final View.OnClickListener mOnItemClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            final OnStationClickListener listener = mOnStationClickListener;
            if (listener != null) {
                ViewHolder viewHolder = (ViewHolder) v.getTag();
                int position = viewHolder.getAdapterPosition();
                if (position != NO_POSITION) {
                    final Location location = mLocations[position];
                    mHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            listener.onStationClicked(location);
                        }
                    });
                }
            }
        }
    };
    @MainThread
    public StationListAdapter(Context context) {
        mHandler = new Handler();
        trainIcon = ThemeHelper.getThemedDrawable(context, R.attr.ic_direction_railway_24dp);
        busIcon = ThemeHelper.getThemedDrawable(context, R.attr.ic_directions_bus_24dp);

    }

    @Override
    public StationListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView = inflater.inflate(R.layout.item_station, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(StationListAdapter.ViewHolder holder, int position) {
        Location location = mLocations[position];
        holder.stationName.setText(location.getName());
        holder.itemView.setTag(holder);
        holder.itemView.setOnClickListener(mOnItemClickListener);
        switch (location.getType()) {
            case TRAIN:
                holder.transportationIcon.setImageResource(trainIcon);
                break;
            case BUS:
                holder.transportationIcon.setImageResource(busIcon);
                break;
            case POI:
            case ADDRESS:
            case UNKNOWN:
            default:
                holder.transportationIcon.setImageDrawable(null);
        }
    }

    @Override
    public void onViewRecycled(ViewHolder holder) {
        super.onViewRecycled(holder);
        holder.itemView.setTag(null);
        holder.itemView.setOnClickListener(null);
        holder.itemView.setClickable(true);
    }

    @MainThread
    public void setLocations(Location[] locations) {
        mLocations = locations;
        notifyDataSetChanged();
    }

    public void setOnStationClickListener(@Nullable OnStationClickListener onStationClickListener) {
        this.mOnStationClickListener = onStationClickListener;
    }

    @Override
    public int getItemCount() {
        return mLocations.length;
    }

    public interface OnStationClickListener {
        void onStationClicked(Location location);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView stationName;
        private final ImageView transportationIcon;

        public ViewHolder(View itemView) {
            super(itemView);
            stationName = (TextView) itemView.findViewById(R.id.stationName);
            transportationIcon = (ImageView) itemView.findViewById(R.id.transportationIcon);
        }
    }
}

File: app/src/main/java/ch/unstable/ost/StationListAdapter.kt
package ch.unstable.ost

import android.content.Context
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.annotation.MainThread
import androidx.recyclerview.widget.RecyclerView
import ch.unstable.ost.api.model.Station
import ch.unstable.ost.api.model.Station.StationType
import ch.unstable.ost.theme.ThemeHelper

internal class StationListAdapter @MainThread constructor(context: Context) : RecyclerView.Adapter<StationListAdapter.ViewHolder>() {
    private val mHandler: Handler = Handler()

    @DrawableRes
    private val trainIcon: Int = ThemeHelper.getThemedDrawable(context, R.attr.ic_direction_railway_24dp)

    @DrawableRes
    private val busIcon: Int = ThemeHelper.getThemedDrawable(context, R.attr.ic_directions_bus_24dp)

    @DrawableRes
    private val tramIcon: Int = ThemeHelper.getThemedDrawable(context, R.attr.ic_direction_tram_24dp)


    var locations = listOf<Station>()
        set(value) {
            field = value
            notifyDataSetChanged()
        }
    private var mOnStationClickListener: OnStationClick? = null
    private val mOnItemClickListener = View.OnClickListener { v ->
        val listener = mOnStationClickListener
        if (listener != null) {
            val viewHolder = v.tag as ViewHolder
            val position = viewHolder.adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                val location = locations[position]
                mHandler.post {
                    listener(location)
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.item_station, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val location = locations[position]
        holder.stationName.text = location.name
        holder.itemView.tag = holder
        holder.itemView.setOnClickListener(mOnItemClickListener)
        when (location.type) {
            StationType.TRAIN -> holder.transportationIcon.setImageResource(trainIcon)
            StationType.BUS -> holder.transportationIcon.setImageResource(busIcon)
            StationType.TRAM -> holder.transportationIcon.setImageResource(tramIcon)
            StationType.POI, StationType.ADDRESS, StationType.UNKNOWN -> holder.transportationIcon.setImageDrawable(null)
            else -> holder.transportationIcon.setImageDrawable(null)
        }
    }

    override fun onViewRecycled(holder: ViewHolder) {
        super.onViewRecycled(holder)
        holder.itemView.tag = null
        holder.itemView.setOnClickListener(null)
        holder.itemView.isClickable = true
    }


    fun setOnStationClickListener(onStationClickListener: OnStationClick) {
        mOnStationClickListener = onStationClickListener
    }

    override fun getItemCount(): Int {
        return locations.size
    }



    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val stationName: TextView = itemView.findViewById(R.id.stationName)
        val transportationIcon: ImageView = itemView.findViewById(R.id.transportationIcon)
    }

}

typealias OnStationClick = (Station) -> Unit
